File: packages/api/aqapi/core/domain/base_collection.py

from abc import ABC, abstractmethod
from typing import Generic, TypeVar, Hashable

from aqapi.core.domain.base_model import BaseModel
from aqapi.core.domain.value_object.base_id import BaseId

IdType = TypeVar("IdType", bound=BaseId)

class CollectionItemProtocol(Generic[IdType]):
    """コレクションアイテムのプロトコル"""
    
    @property
    def id(self) -> IdType:
        """アイテムのIDを返す"""
        raise NotImplementedError("Subclasses must implement this method to return the item ID.")


ItemType = TypeVar("ItemType", bound=BaseModel)

class BaseCollection(ABC, Generic[ItemType, IdType]):

    def __init__(self, items: list[ItemType]):
        self._items = items
        self._item_by_ids: dict[IdType, ItemType] = {}
        self.update_index()

    @classmethod
    def from_empty(cls):
        return cls(items=[])

    def update_index(self):
        """辞書の更新"""
        self._item_by_ids = {item.id: item for item in self._items}
        self._update_custom_index()

    def append(self, item: ItemType) -> None:
        """アイテムを追加"""
        if not isinstance(item, BaseModel):
            raise TypeError(f"Expected item of type {ItemType}, got {type(item)}")
        self._items.append(item)
        self.update_index()

    def get(self, item_id: IdType) -> ItemType | None:
        """IDでアイテムを取得"""
        if not isinstance(item_id, Hashable):
            raise TypeError(f"Expected item_id to be hashable, got {type(item_id)}")
        return self._item_by_ids.get(item_id)

    @abstractmethod
    def _update_custom_index(self) -> None:
        """具象クラスで実装するカスタムインデックスの更新メソッド（任意）"""
        pass

File: packages/backend/src/core/domain/baseCollection.ts
import { BaseModel } from './baseModel';
import { BaseId } from '../../../../shared/core/value-object/base_id';

/**
 * コレクションアイテムのプロトコル（インターフェース）
 */
export interface CollectionItemProtocol<TId extends BaseId> {
  /**
   * アイテムのIDを返す
   */
  readonly id: TId;
}

/**
 * ドメインモデルのコレクションを表現する基底抽象クラス（ファーストコレクション）
 */
export abstract class BaseCollection<
  TItem extends BaseModel<TId, any> & CollectionItemProtocol<TId>,
  TId extends BaseId
> {
  protected _items: TItem[];
  protected _itemByIds: Map<string, TItem> = new Map();

  constructor(items: TItem[]) {
    this._items = items;
    this.updateIndex();
  }

  /**
   * 空のコレクションを作成する
   */
  static fromEmpty<T extends BaseCollection<any, any>>(
    this: new (items: any[]) => T
  ): T {
    return new this([]);
  }

  /**
   * インデックス辞書の更新
   */
  updateIndex(): void {
    this._itemByIds.clear();
    for (const item of this._items) {
      // BaseIdのhash()メソッドを使用してMapのキーにする
      const key = item.id.hash().toString();
      this._itemByIds.set(key, item);
    }
    this._updateCustomIndex();
  }

  /**
   * アイテムを追加
   */
  append(item: TItem): void {
    if (!item || typeof item !== 'object') {
      throw new TypeError(`Expected item of valid type, got ${typeof item}`);
    }
    this._items.push(item);
    this.updateIndex();
  }

  /**
   * IDでアイテムを取得
   */
  get(itemId: TId): TItem | null {
    const key = itemId.hash().toString();
    return this._itemByIds.get(key) || null;
  }

  /**
   * アイテム数を取得
   */
  get length(): number {
    return this._items.length;
  }

  /**
   * すべてのアイテムを取得（読み取り専用）
   */
  get items(): readonly TItem[] {
    return [...this._items];
  }

  /**
   * 具象クラスで実装するカスタムインデックスの更新メソッド（任意）
   */
  protected abstract _updateCustomIndex(): void;
}
